### DISCUSSION TITLE
PSB support

### DISCUSSION THREAD

**Discussion by hyperknot** [libvips/libvips#2002] [Ideas]
Following your [great answer on SO](https://stackoverflow.com/a/44718669/518169), we started using VIPS to process gigapixel images to tiled JPG images. Especially using mozjpeg, the results are really nice and performant.

One problem I have is PSB support. Do I understand right, that PSB support is provided only through ImageMagick, which means that we are limited to loading the whole image to memory?

Is BigTIFF support so efficient because it is using a native TIFF library which has higher degree of memory management option?

Is there any solution / workaround to `dszave` bigger-than-ram PSB images using VIPS?

Would integrating PSD parser be an option? Or to try to convert PSB to (Big) PNG (is there any limit on PNG reading in VIPS) and then read PNG efficiently?
http://telegraphics.com.au/sw/product/psdparse

**Comment by jcupitt** [comment]
Hello, yes, PSB is only supported via MagickCore, so it's very heavy on memory. You could try converting to TIFF or PNG using an external utility, but I don't know of a good one off-hand. 

To add support directly to libvips, we'd need to find a good PSB load library. This looks like the best one:

https://github.com/Code-ReaQtor/libqpsd

But it's not quite right. It would need reworking before it could be used by non-Qt programs, it only seems to support basic PSB images, it's not available on many platforms, and it's not clear if it is being maintained.

**Comment by jcupitt** [comment]
... so the simplest solution would probably be to save as TIFF or PNG from Photoshop, though I realize that is not always possible.

**Comment by hyperknot** [comment]
What do you think about psdparse? It open source, some kind of GPL license. It says it can convert PSB to Big PNG, where Big PNG is something which doesn't have a pixel limitation. But I believe using it directly to read the file would be much faster / more efficient.
http://telegraphics.com.au/sw/product/psdparse

Is there any chance VIPS supports reading these (Big) PNG images produced by psdparse?

**Comment by jcupitt** [comment]
It looks nice, but it's a command-line tool, not a library, so it couldn't be added to vips without quite a lot of work. 

Yes, the PNG files it makes should work well with libvips.

**Comment by hyperknot** [comment]
You are right, I just sought maybe it's structured in a way that the parsing bit can be easily used in other projects. 

libqpsd looks really clean, but it depends on all those Qt things. On the other hand it's only 1000 lines including support for all the edge-cases of the formats, so might not be impossibly hard to pick the PSB relevant bits and try to clean it from Qt-ness.

I understand that it's quite some work to integrate any of those, and I'll have a look at writing to a temporary PNG file.

**Comment by jcupitt** [comment]
Let's leave this open as an enhancement issue.
